contracts/extensions/OneTxPayment.sol

```diff
@@ -181,100 +149,86 @@ contract OneTxPayment is ColonyExtension {
+      uint256 domainPotId = colony.getDomain(_domainId).fundingPotId;
+
+      uint256 idx;
+      uint256 slot;
+
+      for (idx = 0; idx < _workers.length; idx++) {
+        moveFundsWithinDomain(_permissionDomainId, _childSkillIndex, domainPotId, fundingPotId, _amounts[idx], _tokens[idx]);
+
+         // If a new worker, start a new slot
+        if (idx == 0 || _workers[idx] != _workers[idx-1]) {
+          require(idx == 0 || _workers[idx] > _workers[idx-1], "one-tx-payment-bad-worker-order");
+
+          slot++;
+          colony.setExpenditureRecipient(expenditureId, slot, _workers[idx]);
+
+          if (_skillId != 0) {
+            colony.setExpenditureSkill(expenditureId, slot, _skillId);
           }
-        colony.setExpenditurePayout(paymentData[1], paymentData[0], _tokens[index], _amounts[index]);
-      }
-      // Fund the payment
-      colony.moveFundsBetweenPots(
-        _permissionDomainId,
-        _childSkillIndex,
-        _childSkillIndex,
-        domain.fundingPotId,
-        paymentData[2],
-        _amounts[index],
-        _tokens[index]
-      );
-    }
-    if (_tokens.length > 1) {
-      colony.finalizeExpenditure(paymentData[1]);
-      paymentData[0] = 0;
-      // Claim payout on behalf of the recipients
-      for (uint256 index = 0; index < _workers.length; index++) {
-        if (index != 0 && _workers[index] != _workers[index-1]) {
-          paymentData[0]++;
+        } else {
+          require(_tokens[idx] > _tokens[idx-1], "one-tx-payment-bad-token-order");
         }
-        colony.claimExpenditurePayout(paymentData[1], paymentData[0], _tokens[index]);
+
+        colony.setExpenditurePayout(expenditureId, slot, _tokens[idx], _amounts[idx]);
       }
-    } else {
-      colony.finalizePayment(_permissionDomainId, _childSkillIndex, paymentData[1]);
-      // Claim payout on behalf of the recipient
-      colony.claimPayment(paymentData[1], _tokens[0]);
+
+      finalizeAndClaim(expenditureId, _workers, _tokens);
+
     }
   }
 
-  function validateCallerPermissions(
-    uint256 _callerPermissionDomainId,
-    uint256 _callerChildSkillIndex,
-    uint256 _domainId) internal view
+  function moveFundsWithinDomain(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _domainPotId,
+    uint256 _fundingPotId,
+    uint256 _amount,
+    address _token
+  )
+    internal
   {
-    require(
-      ColonyAuthority(colony.authority()).canCall(msg.sender, _callerPermissionDomainId, address(colony), ADD_PAYMENT_SIG),
-      "colony-one-tx-payment-administration-not-authorized"
-    );
-    require(
-      ColonyAuthority(colony.authority()).canCall(msg.sender, _callerPermissionDomainId, address(colony), MOVE_FUNDS_SIG),
-      "colony-one-tx-payment-funding-not-authorized"
-    );
+    colony.moveFundsBetweenPots(_permissionDomainId, _childSkillIndex, _childSkillIndex, _domainPotId, _fundingPotId, _amount, _token);
```

Comment:
There was some reason I did this... although it might have been to avoid putting all the arguments on their own lines, which I suppose isn't a great reason.

Reply:
Nope, it was because of a stack depth error.

Reply:
Ooof, that's annoying. Maybe leave a comment explaining that's what this is working around then so we don't find ourselves wondering in the future?